### DISCUSSION TITLE
courses/computer-structure/%EC%9E%90%EB%A3%8C%EA%B5%AC%EC%A1%B0/%EC%97%B0%EA%B2%B0%EB%A6%AC%EC%8A%A4%ED%8A%B8

### DISCUSSION THREAD

**Discussion by giscus** [crwdu/lms-contents-js101#73] [Q&A]
# courses/computer-structure/%EC%9E%90%EB%A3%8C%EA%B5%AC%EC%A1%B0/%EC%97%B0%EA%B2%B0%EB%A6%AC%EC%8A%A4%ED%8A%B8

https://lms.vanillacoding.co/courses/computer-structure/%EC%9E%90%EB%A3%8C%EA%B5%AC%EC%A1%B0/%EC%97%B0%EA%B2%B0%EB%A6%AC%EC%8A%A4%ED%8A%B8

**Comment by kam1989** [comment]
1: 의사코드 작성

순환만 체크하는 함수(연결리스트) {

  이전 노드들 집합

  연결리스트 순회{
    if 노드.next == null(혹은 끝 등등){
      return success
    }
    else if 이전 노드들.contains(노드.next){
      return error
    }
	
    이전 노드들.add(now)
  }
}

**Comment by kam1989** [comment]
2: 실제코드 작성(의사코드와 다른 방법으로 구현)

function createNode(value) {
  return {
    data: value,
    next: null
  };
}

function createLinkedList() {
  const linkedList = {
    head: null,
  };

  linkedList.addToHead = function (value) {
    const newNode = createNode(value);
    newNode.next = linkedList.head;
    linkedList.head = newNode;
  };

  linkedList.contains = function (value) {
    let current = linkedList.head;

    while (current !== null) {
      if (current.data === value) {
        return true;
      }

      current = current.next;
    }

    return false;
  };

  return linkedList;
}

function checkCirculer(dataList){
    let slowNode = dataList.head;
    let fastNode = dataList.head;
    
    while(fastNode.next !== null){
        fastNode = fastNode.next;
        slowNode = slowNode.next;

        if(fastNode.next !== null){
            fastNode = fastNode.next;
            
            console.log("slowData(current.next)=" + slowNode.data);
            console.log("fastData(current.next.next)=" + fastNode.data);
            console.log("");
            
            if(slowNode === fastNode){
                return false;
            }
        }
    }
    
    return true;
}

//의도적으로 next 링크를 꼬아서 순환 링크드리스트 생성
//circleStart > 1 이면 순환 전에 생성된 노드들은 접근 불가능. 메모리 릭
//circleStart == 1 이면 처음부터 끝까지 하나로 순환되는 구조로 생성된다.
//첫 변수 <= 두번째 변수 <= 세번째 변수
function createCircularList(circleStart, circleEnd, maxLength){
    const linkedList = createLinkedList();
    let node;
    let i;

    if(circleStart == 0){
        console.log("작업실패: 리스트 초기값은 null 입니다.")
        return null;
    }

    //순환이 시작되는 노드 까지 입력 후 노드 백업
    for (i = 0; i < circleStart; i++){
        linkedList.addToHead(i);
    }
    node = linkedList.head;

    //순환이 끝나는 노드까지 입력 후 순환 연결
    for (i; i < circleEnd; i++){
        linkedList.addToHead(i);
    }
    node.next = linkedList.head;

    //나머지 입력
    for (i; i < maxLength; i++){
        linkedList.addToHead(i);
    }

    //확인용 출력
    node = linkedList.head;
    
    for(i = 0; i < maxLength; i++){
        console.log(node.data);
        node = node.next;
    }

    return linkedList;
}

const normalList = createLinkedList();
normalList.addToHead(0);
normalList.addToHead(1);
normalList.addToHead(2);
console.log("순환여부 체크 시작");
console.log(checkCirculer(normalList));
console.log("");

console.log("테스트용 순환배열 생성")
const testList = createCircularList(8,17,20);
console.log("");
console.log("순환여부 체크 시작");
console.log(checkCirculer(testList));

**Comment by smileboy0014** [comment]
1. 의사코드 작성

1. 노드의 헤드 정보를 받는다.
2. 만약에 헤드가 null이면 false를 리턴한다.
3. 순환을 시작한다.
 3.1 헤드가 null이 아니면서 헤드의 next가 null이 아니면 순환을 계속한다.
 3.2 만약 이전 노드.next(1) 와 이전 노드의.next.next(2)가 같은 노드면 true를 리턴한다.
 3.3 비교 노드 정보를 1번과 2번의 노드로 갱신한다.
 3.4 순환을 계속한다.
4. false를 리턴한다.

**Comment by cholulu99** [comment]
check함수는 current.next주소값과 같은 주소값이 있는지 확인하는 코드
체크함수는 두개 매개변수 ( 주소값, 인덱스)

check함수 안에는 index만큼 반복하는 반복문이 있음
head부터 current 전까지 주소값을 확인하기 위함
인덱스가 1일땐 head 까지만 확인
인덱스가 2일때는 head와 head 다음 노드 까지 확인



current = head 
index = 0;
while(current!==NULL)
current = current.next
index++
check(current.next,index)

이런 양식으로 하는게 맞는지 몰라서 일단 적어봅니다 ... ㅎ
